### PR TITLE
Converter for functor system

### PR DESCRIPTION
**Pull request by GiudGiud**
anticipated the disappearance of declareFunctorProperty and just called it FunctorADConverter instead of FunctorMaterialConverter

### CODE REVIEW

File: framework/src/materials/FunctorADConverter.C [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  params.addParam<std::vector<std::string>>("reg_props_out",
+                                            "The names of the output regular functors");
+  return params;
+}
+
+template <typename T>
+FunctorADConverterTempl<T>::FunctorADConverterTempl(const InputParameters & parameters)
+  : FunctorMaterial(parameters)
+{
+  auto reg_props_in = getParam<std::vector<std::string>>("reg_props_in");
```
Comment (dschwen) [inline framework/src/materials/FunctorADConverter.C:38]: It'd be more elegant to use `getParam<std::string, std::string>("reg_props_in", "ad_props_out");` to have MOOSE build a vector of pairs

Comment (GiudGiud) [reply framework/src/materials/FunctorADConverter.C:38]: could do a map too, but all the Generic Materials are not built this way

**Comment by GiudGiud** [comment]
Regular ADConverter, which I've only renamed, is covered in tensor_mechanics. I think there's a case for merging with poor coverage haha

**Comment by GiudGiud** [comment]
Ok added a within framework test for materialAdConverter. 
Also should get better coverage for the piecewiseConstantByBlockMaterial now

File: framework/doc/content/source/materials/FunctorADConverter.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+Some examples of safe conversions:
+
+- Functions to ADFunctions and vice-versa are safe since functions do not hold derivative data
+
+- AuxVariables to regular functors is safe since auxiliary variables do not hold derivative data
+
+
+## Example input syntax
+
+An example of some gymnastic with functor conversions is shown in this example. The reader should note
```
Comment (lindsayad) [inline framework/doc/content/source/materials/FunctorADConverter.md:19]:
```suggestion
An example of some gymnastics with functor conversions is shown in this example. The reader should note
```

File: framework/src/materials/FunctorADConverter.C [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+registerMooseObject("MooseApp", FunctorADConverter);
+registerMooseObject("MooseApp", VectorFunctorADConverter);
+
+template <typename T>
+InputParameters
+FunctorADConverterTempl<T>::validParams()
+{
+  InputParameters params = FunctorMaterial::validParams();
+  params.addClassDescription("Converts regular functors to AD functors and "
+                             " AD functors to regular functors");
```
Comment (lindsayad) [inline framework/src/materials/FunctorADConverter.C:23]: seems like an unnecessary space here?

File: framework/src/materials/FunctorADConverter.C [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  // Define the AD functors
+  for (const auto i : make_range(reg_props_in.size()))
+  {
+    const auto & reg_functor = &getFunctor<T>(reg_props_in[i]);
+    const auto & ad_prop =
+        &declareFunctorProperty<typename Moose::ADType<T>::type>(ad_props_out[i]);
+    ad_prop->setFunctor(_mesh,
+                        blockIDs(),
+                        [reg_functor](const auto & r, const auto & t) ->
+                        typename Moose::ADType<T>::type { return (*reg_functor)(r, t); });
```
Comment (lindsayad) [inline framework/src/materials/FunctorADConverter.C:92]:
```suggestion
    const auto & reg_functor = getFunctor<T>(reg_props_in[i]);
    auto & ad_prop =
        declareFunctorProperty<typename Moose::ADType<T>::type>(ad_props_out[i]);
    ad_prop.setFunctor(_mesh,
                        blockIDs(),
                        [&reg_functor](const auto & r, const auto & t) ->
                        typename Moose::ADType<T>::type { return reg_functor(r, t); });
```

File: framework/src/materials/FunctorADConverter.C [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+  }
+
+  // Define the regular functors
+  for (const auto i : make_range(ad_props_in.size()))
+  {
+    const auto & ad_functor = &getFunctor<typename Moose::ADType<T>::type>(ad_props_in[i]);
+    const auto & reg_prop = &declareFunctorProperty<T>(reg_props_out[i]);
+    reg_prop->setFunctor(_mesh, blockIDs(), [ad_functor](const auto & r, const auto & t) -> T {
+      return MetaPhysicL::raw_value((*ad_functor)(r, t));
```
Comment (lindsayad) [inline framework/src/materials/FunctorADConverter.C:101]:
```suggestion
    const auto & ad_functor = getFunctor<typename Moose::ADType<T>::type>(ad_props_in[i]);
    auto & reg_prop = declareFunctorProperty<T>(reg_props_out[i]);
    reg_prop.setFunctor(_mesh, blockIDs(), [&ad_functor](const auto & r, const auto & t) -> T {
      return MetaPhysicL::raw_value(ad_functor(r, t));
```

**Review by lindsayad — CHANGES_REQUESTED** [review]
Why do we even want `PiecewiseConstantByBlockMaterial` any more? It was previously converted to be a functor material so any users of that material should already be accepting functors instead of regular material properties, so it seems like anyone previously using it should just use `PicewiseByBlockFunctorMaterial` instead and that we should consequently have no current users of a regular material version of it.

**Comment by GiudGiud** [comment]
We have a potential user in a discussions thread
In general it s a good idea to have this material, it s a shorter syntax than the existing options
https://github.com/idaholab/moose/discussions/19470

File: framework/src/materials/FunctorADConverter.C (hunk 3 above)
Comment (GiudGiud) [inline framework/src/materials/FunctorADConverter.C:23]:
```suggestion
                             "AD functors to regular functors");
```

**Review by lindsayad — APPROVED** [review]

**Pull request merged**